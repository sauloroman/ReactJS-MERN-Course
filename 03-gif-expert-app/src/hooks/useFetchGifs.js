import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs"

export const useFetchGifs = ( search ) => {

  const [result, setResult] = useState({
    data: null,
    hasError: null,
    isLoading: false,
  })

  useEffect(() => {
    getData( search )
  }, [])

  const getData = async () => {

    setResult({
      ...result,
      isLoading: true,
    })

    try {

      const data = await getGifs( search )

      setResult({
        data,
        hasError: null,
        isLoading: false,
      })
      
    } catch (error) {
      setResult({
        data: null,
        hasError: error,
        isLoading: false,
      })
    }

  }

  return {
    ...result,
  }

}