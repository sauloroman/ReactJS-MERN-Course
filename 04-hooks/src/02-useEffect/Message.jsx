import { useEffect, useState } from "react"

export const Message = () => {
  
  const [coords, setCoords] = useState({ x: 0, y: 0 })

  useEffect(() => {

    const onMouseMove = (event) => {
      const { clientX, clientY } = event
      setCoords({ x: clientX, y: clientY })
    }

    window.addEventListener('mousemove', onMouseMove )

    return () => {
      window.removeEventListener('mousemove', onMouseMove )
    }

  }, [])

  return (
    <p>Coordenadas: <span data-testid="coords">{ JSON.stringify( coords ) }</span></p>
  )
}
