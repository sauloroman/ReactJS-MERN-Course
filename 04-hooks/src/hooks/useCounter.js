import { useState } from "react"

export const useCounter = ( initialValue = 1 ) => {

  const [counter, setCounter] = useState(initialValue)

  const onIncrementCounter = ( value = 1 ) => {
    setCounter( counter => counter + value )
  }

  const onDecrementCounter = ( value = 1 ) => {
    if ( counter === 0 ) return
    setCounter( counter => counter - value )
  }

  const onResetCounter = () => setCounter(initialValue)

  return {
    // State
    counter,

    // Methods
    onIncrementCounter,
    onDecrementCounter,
    onResetCounter
  }

}