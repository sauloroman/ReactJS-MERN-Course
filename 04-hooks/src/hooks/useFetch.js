import { useEffect } from "react"
import { useState } from "react"

export const useFetch = ( url = '' ) => {

  const [state, setState] = useState({
    data: null,
    hasError: null,
    isLoading: false,
  })

  useEffect(() => {
    getData()
  }, [ url ])

  const getData = async () => {

    setState({
      ...state,
      isLoading: true,
    })

    try {
      
      const res = await fetch( url );
      const data = await res.json()
    
      setState({
        ...state,
        data: data,
        isLoading: false,
      })

    } catch (error) {
      setState({
        ...state,
        hasError: error,
        isLoading: false,
      })
    }

  }

  return {
    data: state.data,
    hasError: state.hasError,
    isLoading: state.isLoading,
  }

}