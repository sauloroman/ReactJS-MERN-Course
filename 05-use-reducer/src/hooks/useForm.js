import { useState } from "react"

export const useForm = ( initialState = {} ) => {

  const [formState, setFormState] = useState(initialState)

  const onInputChange = ({ target: { name, value }}) => {
    setFormState({
      ...formState,
      [name]: value
    })
  }

  const onResetForm = () => setFormState( initialState )

  return {
    formState,
    ...formState,

    onInputChange,
    onResetForm,
  }

}