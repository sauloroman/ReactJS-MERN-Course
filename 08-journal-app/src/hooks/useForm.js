import { useEffect, useMemo, useState } from "react"

export const useForm = ( initialState = {}, formValidations = {} ) => {

  const [ formState, setFormState ] = useState( initialState )
  const [formValidation, setFormValidation] = useState({})

  useEffect(() => {
    createValidators()
  }, [formState])

  useEffect(() => {
    setFormState( initialState )
  }, [initialState])

  const isFormValid = useMemo(() => {
    for ( const formValue of Object.values(formValidation) ) {
      if ( formValue !== null ) return false
    }

    return true
  }, [formValidation])

  const onInputChange = ({ target }) => {

    const { name, value } = target

    setFormState({
      ...formState,
      [name]: value
    })

  }

  const onResetForm = () => setFormState( initialState )
  
  const createValidators = () => {

    const formCheckedValues = {}

    for( const formField of Object.keys( formValidations )) {
      const [ fn, errorMessage ] = formValidations[ formField ]
      formCheckedValues[`${formField}Valid`] = fn( formState[formField] ) ? null : errorMessage
    }
 
    setFormValidation({ ...formCheckedValues })

  }

  return {
    ...formState,
    ...formValidation,
    formState,
    isFormValid,

    onInputChange,
    onResetForm
  }

}