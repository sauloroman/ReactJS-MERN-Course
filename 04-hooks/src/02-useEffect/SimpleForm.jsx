import { useEffect, useState } from "react"
import { Message } from "./Message"

export const SimpleForm = () => {
  
  const [formState, setFormState] = useState({
    username: '',
    email: '',
    age: '',
  })
  
  const { username, email, age } = formState

  const onInputChange = ({ target: { name, value }}) => {
    setFormState({
      ...formState,
      [name]: value
    })
  }

  // useEffect(() => {
  //   console.log('The component is mounted')
  // }, [])

  // useEffect(() => {
  //   console.log('Username has changed')
  // }, [username])

  // useEffect(() => {
  //   console.log('Email has changed')
  // }, [email])

  // useEffect(() => {
  //   console.log('Age has changed')
  // }, [age])

  return (
    <>
      <h1 className="heading">UseEffect &mdash; Formulario Simple</h1>

      <form action="#" className="form">
        <div className="form__field">
          <label htmlFor="username" className="form__label">Nombre del usuario</label>
          <input 
            name="username"
            value={ username }
            onChange={onInputChange}
            type="text" 
            className="form__input" 
            placeholder="Ingrese su nombre"
          />
        </div>
        <div className="form__field">
          <label htmlFor="email" className="form__label">Correo Electr√≥nico</label>
          <input 
            name="email"
            value={ email }
            onChange={onInputChange}
            type="email" 
            className="form__input" 
            placeholder="Ingrese su correo"
          />
        </div>
        <div className="form__field">
          <label htmlFor="age" className="form__label">Edad</label>
          <input 
            name="age"
            value={ age }
            onChange={onInputChange}
            type="number"
            min={1} 
            className="form__input" 
            placeholder="Ingrese su edad"
          />
        </div>
      </form>

      { username === 'Roman' && <Message /> }
    </>
  )
}
