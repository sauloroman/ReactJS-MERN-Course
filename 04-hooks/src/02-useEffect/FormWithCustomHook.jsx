import { useForm } from "../hooks/useForm"

export const FormWithCustomHook = () => {

  const { onInputChange, onResetForm, username, email } = useForm({ username: '', email: '' })

  return (
    <>
      <h2 className="heading">UseEffect &mdash; FormCustomHook</h2>

      <form action="#" className="form">
        <div className="form__field">
          <label htmlFor="username" className="form__label">Nombre del usuario</label>
          <input 
            data-testid="username-input"
            id="username"
            name="username"
            value={username}
            onChange={ onInputChange }
            className="form__input" 
            placeholder="Ingrese su nombre" 
            type="text" 
          />
        </div>
        <div className="form__field">
          <label htmlFor="email" className="form__label">Correo del usuario</label>
          <input 
            data-testid="email-input"
            id="email"
            name="email"
            value={email}
            onChange={ onInputChange }
            className="form__input" 
            placeholder="Ingrese su correo" 
            type="text" 
          />
        </div>

        <button onClick={ onResetForm } className="form__btn">Limpiar formulario</button>
      </form>
    </>
  )
}
