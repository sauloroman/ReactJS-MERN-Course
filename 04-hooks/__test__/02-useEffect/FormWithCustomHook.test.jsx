import { fireEvent, render, screen } from "@testing-library/react"
import { FormWithCustomHook } from "../../src/02-useEffect/FormWithCustomHook"
import { useForm } from "../../src/hooks/useForm"

jest.mock('../../src/hooks/useForm.js')

describe('Tests in FormWithCustomHook', () => {

  const onInputChangeMock = jest.fn()
  const onResetFormMock = jest.fn()
  const formStateMock = {
    username: '',
    email: ''
  }

  useForm.mockReturnValue({
    onInputChange: onInputChangeMock,
    onResetForm: onResetFormMock,
    formState: formStateMock, 
  })

  test('it must match with snapshot', () => {
    const { container } = render(<FormWithCustomHook />)
    expect( container ).toMatchSnapshot()
  })

  test('it should call onInputChange function when input changed', () => {
    render(<FormWithCustomHook />)

    const usernameInput = screen.getByTestId('username-input')
    fireEvent.change( usernameInput, { target: {value: 'hello world'} })

    expect( onInputChangeMock ).toHaveBeenCalled()
  })

  test('it should call the onResetForm function when button is pushed', () => {
    render(<FormWithCustomHook />)

    const buttonReset = screen.getByText('Limpiar formulario')
    fireEvent.click( buttonReset )

    expect( onResetFormMock ).toHaveBeenCalled()
  })

})