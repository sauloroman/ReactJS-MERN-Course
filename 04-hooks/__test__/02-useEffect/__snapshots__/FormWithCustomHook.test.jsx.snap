// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in FormWithCustomHook it must match with snapshot 1`] = `
<div>
  <h2
    class="heading"
  >
    UseEffect â€” FormCustomHook
  </h2>
  <form
    action="#"
    class="form"
  >
    <div
      class="form__field"
    >
      <label
        class="form__label"
        for="username"
      >
        Nombre del usuario
      </label>
      <input
        class="form__input"
        data-testid="username-input"
        id="username"
        name="username"
        placeholder="Ingrese su nombre"
        type="text"
        value=""
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="form__label"
        for="email"
      >
        Correo del usuario
      </label>
      <input
        class="form__input"
        data-testid="email-input"
        id="email"
        name="email"
        placeholder="Ingrese su correo"
        type="text"
        value=""
      />
    </div>
    <button
      class="form__btn"
    >
      Limpiar formulario
    </button>
  </form>
</div>
`;
